### PR TITLE
Switch from TSLint to ESLint

### PR DESCRIPTION
**Pull request by gabebw**
The bulk of this work was done automatically by `npx tslint-to-eslint-config` according to the instructions [here](https://github.com/typescript-eslint/tslint-to-eslint-config#tslint-to-eslint-config).

I then manually compared the ESLint rules to the TSLint rules. Then I ran ESLint over the existing files and fixed rules that did not error but should (false negative).

I also copied lines that were ignored by a TSLint comment to also be ignored by an ESLint comment.

Each commit here is self-contained, so I recommend going through each commit one by one to see the process. The commits deal with, in rough order from oldest to newest:

* Upgrade TSLint to a version that works with `tslint-to-eslint-config`
* Generate ESLint config
* Remove TSLint and related packages from `package.json`
* Fix linting errors (the error output is listed in the commit messages)

Because ESLint is not recursive in the same way TSLint is, and does not typecheck, each package has its own `yarn lint` task that runs ESLint and the typescript compiler (for typechecking).

To run all linting tasks for all packages, run `yarn lint` in the root directory.

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-v8.11.3
\ No newline at end of file
+v12.14.1
```
Comment (gabebw) [inline .nvmrc:1]: `npx tslint-to-eslint-config` needed a new version of Node, so I picked this one.

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            "always",
+            {
+                "markers": [
+                    "/"
+                ]
+            }
+        ],
+        "use-isnan": "error",
+        /*
+        "@typescript-eslint/tslint/config": [
```
Comment (gabebw) [inline .eslintrc.js:189]: This is a list of rules from TSLint that were not automatically copied over to ESLint.

I think the biggest one is probably `function-name`, which specifies the format for private/static/function methods (like `__private_method`). I don't know how to make that happen in ESLint.

**Comment by blacktaxi** [comment]
Hey @oleksiilevzhynskyi, thanks for keeping this up to date!

I noticed that the `no-redeclare` rule seems to trigger a lot in this codebase, since we use the type + namespace merging pattern a lot. I tried to use a TS-specific alternative (`@typescript-eslint/no-redeclare`), but it actually [intentionally flags such use case](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-redeclare.md).

I see this as a useful lint rule overall, but also don't really agree on the default behavior. Given how both `no-redeclare` and `@typescript-eslint/no-redeclare` basically give the same result in our use case, it may seem that there's no difference but I think it's probably still better to use the TS-specific version.

Thoughts?

**Comment by oleksiilevzhynskyi** [comment]
@blacktaxi, thanks for pointing this out. `@typescript-eslint/no-redeclare` complaining only about the "merge" declaration of Atom. Another case it treats as correct usage. Changed.

File: .eslintrc.js (hunk 2 above)
Comment (oleksiilevzhynskyi) [reply .eslintrc.js:202]: There is no replacement yet. I think we're ok to go without this rule.

**Comment by blacktaxi** [comment]
Thanks! Everything is green, but I checked the CI log and it seems that eslint doesn't work as expected on the tests and examples code ([see here](https://github.com/grammarly/focal/runs/1742248559#step:6:222)):

```
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/app.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/add-input/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/animated-counter/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/big-table/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/bmi/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/change-color/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
Tried to lint /home/runner/work/focal/focal/packages/examples/all/src/checkbox-undo-redo/index.tsx but found no valid, enabled rules for this file type and file path in the resolved configuration.
...
```

When I run this locally, I actually get a bunch of lint errors in these files (basically every time `AppState` is declared).

**Comment by oleksiilevzhynskyi** [comment]
Migrated to [eslint-webpack-plugin](https://github.com/webpack-contrib/eslint-webpack-plugin) from ts-loader.

[tslint-loader](https://github.com/wbuchwalter/tslint-loader#tslint-loader) was deprecated in favor of eslint-loader; [eslint-loader](https://github.com/webpack-contrib/eslint-loader#warning-deprecated) was deprecated in favor of **eslint-webpack-plugin**

**Review by blacktaxi — APPROVED** [review]

**Comment by oleksiilevzhynskyi** [comment]
@gabebw thank you for MR. eslint is finally merged! 🎉

**Comment by gabebw** [comment]
Thank you for getting it merged!

**Pull request merged**